In [1]:
import pandas as pd

In [2]:
df = pd.read_csv("SBAnational.csv")

/tmp/ipykernel_183787/3536806131.py:1: DtypeWarning: Columns (9) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("SBAnational.csv")


## NewExist

In [51]:
df["NewExist"].describe()

count    899164.00000
mean          1.28021
std           0.45199
min           0.00000
25%           1.00000
50%           1.00000
75%           2.00000
max           2.00000
Name: NewExist, dtype: float64

In [52]:
df.loc[df["NewExist"].isnull(), "NewExist"]

Series([], Name: NewExist, dtype: float64)

In [53]:
print(df.loc[df["NewExist"].isnull(), ["ApprovalFY"]])

Empty DataFrame
Columns: [ApprovalFY]
Index: []


In [ ]:
print(df.loc[df["NewExist"].isnull(), ["ApprovalFY"]].sort_values(by="ApprovalFY", ascending=True))

NewExist (1 D Existing Business, 2 D New Business)

In [92]:
df["NewExist"].isna().any()

np.False_

In [98]:
df.query("NewExist == 0")

LoanNr_ChkDgt                            Name               City  \
142        1001235008  NEW ENGLAND PAINT & SUPPLY, IN  READING (TOWN OF)   
170        1001385007                  Mill City Club           Lawrence   
185        1001485002                LBC BILLING, LLC         PROVIDENCE   
211        1001685003  Comprehensive Psychiatric Asso   Weston (Town of)   
223        1001785009     West Reading Used Furniture       West Reading   
...               ...                             ...                ...   
895872     9923063005     VAN-JONES HARDWOOD CO., INC     CRYSTAL SPRING   
895976     9925253007           ACO ENTERPRISES, INC.     SALT LAKE CITY   
896382     9933633003           FIRST CLASS TAN, INC.        HOT SPRINGS   
896633     9939223010   BO SHANNON GRAPHIC PRODUCTION       LEE'S SUMMIT   
897212     9952683002     CLEAN & DRY LAUNDRY SERVICE              BUTTE   

       State    Zip                         Bank BankState   NAICS  \
142       MA   1867                  DANVERSBANK        MA       0   
170       MA   1841   BANK OF AMERICA NATL ASSOC        NC  722410   
185       RI   2908   BANK OF AMERICA NATL ASSOC        NC  514210   
211       MA   2193   BANK OF AMERICA NATL ASSOC        NC       0   
223       PA  19611  WELLS FARGO BANK NATL ASSOC        PA       0   
...      ...    ...                          ...       ...     ...   
895872    PA  15536    FARMERS & MERCHANTS TR CO        PA  421310   
895976    UT  84050  WELLS FARGO BANK NATL ASSOC        SD  316991   
896382    AR  71913   BANK OF AMERICA NATL ASSOC        AR       0   
896633    MO  64064            COUNTRY CLUB BANK        KS       0   
897212    MT  59701     FIRST MONTANA BANK, INC.        MT  812310   

       ApprovalDate ApprovalFY  ...  RevLineCr  LowDoc  ChgOffDate  \
142       21-May-97       1997  ...          0       N         NaN   
170       16-Jun-97       1997  ...          0       N         NaN   
185        2-Jul-97       1997  ...          0       N         NaN   
211       19-Sep-97       1997  ...          0       N         NaN   
223       10-Oct-97       1998  ...          0       N         NaN   
...             ...        ...  ...        ...     ...         ...   
895872    27-Jan-97       1997  ...          Y       N         NaN   
895976    28-Jan-97       1997  ...          N       N         NaN   
896382    30-Jan-97       1997  ...          N       Y         NaN   
896633     3-Feb-97       1997  ...          N       Y         NaN   
897212    10-Feb-97       1997  ...          N       N         NaN   

        DisbursementDate  DisbursementGross  BalanceGross  MIS_Status  \
142            31-Mar-99       $100,000.00         $0.00        P I F   
170            30-Nov-97        $70,000.00         $0.00        P I F   
185            30-Apr-99        $10,000.00         $0.00        P I F   
211            30-Nov-97       $100,000.00         $0.00        P I F   
223            30-Apr-98        $20,000.00         $0.00        P I F   
...                  ...                ...           ...         ...   
895872         28-Feb-97       $250,000.00         $0.00        P I F   
895976         28-Feb-97        $50,000.00         $0.00        P I F   
896382         28-Feb-97        $18,700.00         $0.00        P I F   
896633         30-Jun-97        $60,000.00         $0.00        P I F   
897212         30-Jun-97       $240,000.00         $0.00        P I F   

       ChgOffPrinGr        GrAppv      SBA_Appv  
142          $0.00   $100,000.00    $50,000.00   
170          $0.00    $70,000.00    $35,000.00   
185          $0.00    $10,000.00     $5,000.00   
211          $0.00   $100,000.00    $50,000.00   
223          $0.00    $20,000.00    $10,000.00   
...             ...           ...           ...  
895872       $0.00   $250,000.00   $225,000.00   
895976       $0.00    $50,000.00    $45,000.00   
896382       $0.00    $23,000.00    $18,400.00   
896633       $0.00    $60,000.00    

In [101]:
df.query("NewExist == 1")

LoanNr_ChkDgt                            Name             City State  \
2          1000034009           WHITLOCK DDS, TODD M.      BLOOMINGTON    IN   
3          1000044001   BIG BUCKS PAWN & JEWELRY, LLC     BROKEN ARROW    OK   
4          1000054004     ANASTASIA CONFECTIONS, INC.          ORLANDO    FL   
5          1000084002  B&T SCREW MACHINE COMPANY, INC       PLAINVILLE    CT   
11         1000144007  PROFESSIONAL ELEVATOR SERVICES          CHICAGO    IL   
...               ...                             ...              ...   ...   
899157     9995493004                  PULLTARPS MFG.         EL CAJON    CA   
899159     9995573004                    FABRIC FARMS  UPPER ARLINGTON    OH   
899160     9995603000                    FABRIC FARMS         COLUMBUS    OH   
899161     9995613003    RADCO MANUFACTURING CO.,INC.      SANTA MARIA    CA   
899162     9995973006           MARUTAMA HAWAII, INC.         HONOLULU    HI   

          Zip                            Bank BankState   NAICS ApprovalDate  \
2       47401         GRANT COUNTY STATE BANK        IN  621210    28-Feb-97   
3       74012   1ST NATL BK & TR CO OF BROKEN        OK       0    28-Feb-97   
4       32801         FLORIDA BUS. DEVEL CORP        FL       0    28-Feb-97   
5        6062   TD BANK, NATIONAL ASSOCIATION        DE  332721    28-Feb-97   
11      60605      BANK OF AMERICA NATL ASSOC        OR  235950    28-Feb-97   
...       ...                             ...       ...     ...          ...   
899157  92020  U.S. BANK NATIONAL ASSOCIATION        CA  314912    27-Feb-97   
899159  43221  JPMORGAN CHASE BANK NATL ASSOC        IL  451120    27-Feb-97   
899160  43221  JPMORGAN CHASE BANK NATL ASSOC        IL  451130    27-Feb-97   
899161  93455  RABOBANK, NATIONAL ASSOCIATION        CA  332321    27-Feb-97   
899162  96830                  BANK OF HAWAII        HI       0    27-Feb-97   

       ApprovalFY  ...  RevLineCr  LowDoc  ChgOffDate  DisbursementDate  \
2            1997  ...          N       N         NaN         31-Dec-97   
3            1997  ...          N       Y         NaN         30-Jun-97   
4            1997  ...          N       N         NaN         14-May-97   
5            1997  ...          N       N         NaN         30-Jun-97   
11           1997  ...          N       N         NaN         30-Nov-97   
...           ...  ...        ...     ...         ...               ...   
899157       1997  ...          N       N         NaN         31-Mar-97   
899159       1997  ...          0       N         NaN         30-Sep-97   
899160       1997  ...          Y       N         NaN         31-Oct-97   
899161       1997  ...          N       N         NaN         30-Sep-97   
899162       1997  ...          N       Y    8-Mar-00         31-Mar-97   

        DisbursementGross  BalanceGross  MIS_Status ChgOffPrinGr  \
2            $287,000.00         $0.00        P I F       $0.00    
3             $35,000.00         $0.00        P I F       $0.00    
4            $229,000.00         $0.00        P I F       $0.00    
5            $517,000.00         $0.00        P I F       $0.00    
11           $150,000.00         $0.00        P I F       $0.00    
...                   ...           ...         ...          ...   
899157       $200,000.00         $0.00        P I F       $0.00    
899159        $70,000.00         $0.00        P I F       $0.00    
899160        $85,000.00         $0.00        P I F       $0.00    
899161       $300,000.00         $0.00        P I F       $0.00    
899162        $75,000.00         $0.00       CHGOFF  $46,383.00    

              GrAppv      SBA_Appv  
2       $287,000.00   $215,250.00   
3        $35,000.00    $28,000.00   
4       $229,000.00   $229,000.00   
5       $517,000.00   $387,750.00   
11      $300,000.00   $225,000.00   
...              ...           ...  
899157  $200,000.00   $150,000.00   
899159   $70,000.00    $56,000.00   
899160   $85,000.00    $42,500.00  

In [118]:
# df[(df["NewExist"]==2)& (df["ApprovalFY"] = 2005)][['Term']]


df[(df["NewExist"] == 0)&(df["MIS_Status"] == "P I F")][["NoEmp","GrAppv","ApprovalDate"]]



NoEmp        GrAppv ApprovalDate
142         4  $100,000.00     21-May-97
170         2   $70,000.00     16-Jun-97
185         1   $10,000.00      2-Jul-97
211        10  $100,000.00     19-Sep-97
223         3   $20,000.00     10-Oct-97
...       ...           ...          ...
895872      7  $250,000.00     27-Jan-97
895976      2   $50,000.00     28-Jan-97
896382      3   $23,000.00     30-Jan-97
896633      1   $60,000.00      3-Feb-97
897212      3  $240,000.00     10-Feb-97

[1085 rows x 3 columns]

In [120]:
df[(df["NewExist"] == 0)&(df["NoEmp"] != "0")][["NoEmp","GrAppv","ApprovalDate"]]

NoEmp        GrAppv ApprovalDate
142         4  $100,000.00     21-May-97
170         2   $70,000.00     16-Jun-97
185         1   $10,000.00      2-Jul-97
211        10  $100,000.00     19-Sep-97
223         3   $20,000.00     10-Oct-97
...       ...           ...          ...
895872      7  $250,000.00     27-Jan-97
895976      2   $50,000.00     28-Jan-97
896382      3   $23,000.00     30-Jan-97
896633      1   $60,000.00      3-Feb-97
897212      3  $240,000.00     10-Feb-97

[1170 rows x 3 columns]

In [107]:
df[(df["NewExist"] == 0)]

LoanNr_ChkDgt                            Name               City  \
142        1001235008  NEW ENGLAND PAINT & SUPPLY, IN  READING (TOWN OF)   
170        1001385007                  Mill City Club           Lawrence   
185        1001485002                LBC BILLING, LLC         PROVIDENCE   
211        1001685003  Comprehensive Psychiatric Asso   Weston (Town of)   
223        1001785009     West Reading Used Furniture       West Reading   
...               ...                             ...                ...   
895872     9923063005     VAN-JONES HARDWOOD CO., INC     CRYSTAL SPRING   
895976     9925253007           ACO ENTERPRISES, INC.     SALT LAKE CITY   
896382     9933633003           FIRST CLASS TAN, INC.        HOT SPRINGS   
896633     9939223010   BO SHANNON GRAPHIC PRODUCTION       LEE'S SUMMIT   
897212     9952683002     CLEAN & DRY LAUNDRY SERVICE              BUTTE   

       State    Zip                         Bank BankState   NAICS  \
142       MA   1867                  DANVERSBANK        MA       0   
170       MA   1841   BANK OF AMERICA NATL ASSOC        NC  722410   
185       RI   2908   BANK OF AMERICA NATL ASSOC        NC  514210   
211       MA   2193   BANK OF AMERICA NATL ASSOC        NC       0   
223       PA  19611  WELLS FARGO BANK NATL ASSOC        PA       0   
...      ...    ...                          ...       ...     ...   
895872    PA  15536    FARMERS & MERCHANTS TR CO        PA  421310   
895976    UT  84050  WELLS FARGO BANK NATL ASSOC        SD  316991   
896382    AR  71913   BANK OF AMERICA NATL ASSOC        AR       0   
896633    MO  64064            COUNTRY CLUB BANK        KS       0   
897212    MT  59701     FIRST MONTANA BANK, INC.        MT  812310   

       ApprovalDate ApprovalFY  ...  RevLineCr  LowDoc  ChgOffDate  \
142       21-May-97       1997  ...          0       N         NaN   
170       16-Jun-97       1997  ...          0       N         NaN   
185        2-Jul-97       1997  ...          0       N         NaN   
211       19-Sep-97       1997  ...          0       N         NaN   
223       10-Oct-97       1998  ...          0       N         NaN   
...             ...        ...  ...        ...     ...         ...   
895872    27-Jan-97       1997  ...          Y       N         NaN   
895976    28-Jan-97       1997  ...          N       N         NaN   
896382    30-Jan-97       1997  ...          N       Y         NaN   
896633     3-Feb-97       1997  ...          N       Y         NaN   
897212    10-Feb-97       1997  ...          N       N         NaN   

        DisbursementDate  DisbursementGross  BalanceGross  MIS_Status  \
142            31-Mar-99       $100,000.00         $0.00        P I F   
170            30-Nov-97        $70,000.00         $0.00        P I F   
185            30-Apr-99        $10,000.00         $0.00        P I F   
211            30-Nov-97       $100,000.00         $0.00        P I F   
223            30-Apr-98        $20,000.00         $0.00        P I F   
...                  ...                ...           ...         ...   
895872         28-Feb-97       $250,000.00         $0.00        P I F   
895976         28-Feb-97        $50,000.00         $0.00        P I F   
896382         28-Feb-97        $18,700.00         $0.00        P I F   
896633         30-Jun-97        $60,000.00         $0.00        P I F   
897212         30-Jun-97       $240,000.00         $0.00        P I F   

       ChgOffPrinGr        GrAppv      SBA_Appv  
142          $0.00   $100,000.00    $50,000.00   
170          $0.00    $70,000.00    $35,000.00   
185          $0.00    $10,000.00     $5,000.00   
211          $0.00   $100,000.00    $50,000.00   
223          $0.00    $20,000.00    $10,000.00   
...             ...           ...           ...  
895872       $0.00   $250,000.00   $225,000.00   
895976       $0.00    $50,000.00    $45,000.00   
896382       $0.00    $23,000.00    $18,400.00   
896633       $0.00    $60,000.00    

In [71]:
df[(df["ApprovalFY"]==1990)][['NewExist']]

NewExist
364420       1.0
364516       1.0
364517       1.0
364519       2.0
364520       1.0
...          ...
439717       1.0
439719       1.0
439721       1.0
439724       2.0
439726       1.0

[14859 rows x 1 columns]

In [54]:
df.query("NewExist == 2")

LoanNr_ChkDgt                            Name            City State  \
0          1000014003                  ABC HOBBYCRAFT      EVANSVILLE    IN   
1          1000024006     LANDMARK BAR & GRILLE (THE)       NEW PARIS    IN   
6          1000093009   MIDDLE ATLANTIC SPORTS CO INC           UNION    NJ   
7          1000094005                 WEAVER PRODUCTS     SUMMERFIELD    FL   
8          1000104006                TURTLE BEACH INN  PORT SAINT JOE    FL   
...               ...                             ...             ...   ...   
899133     9995063004    CONTINENTAL CONTAINERS, INC.       CLEVELAND    OH   
899143     9995213001   CITY CONSTRUCTION CORPORATION         MARLTON    NJ   
899151     9995393009  FUTURE CONTRACTING SERVICES, I          DALLAS    PA   
899158     9995563001  SHADES WINDOW TINTING AUTO ALA          IRVING    TX   
899163     9996003010  PACIFIC TRADEWINDS FAN & LIGHT          KAILUA    HI   

          Zip                            Bank BankState   NAICS ApprovalDate  \
0       47711                FIFTH THIRD BANK        OH  451120    28-Feb-97   
1       46526                 1ST SOURCE BANK        IN  722410    28-Feb-97   
6        7083     WELLS FARGO BANK NATL ASSOC        SD       0     2-Jun-80   
7       34491                    REGIONS BANK        AL  811118    28-Feb-97   
8       32456                 CENTENNIAL BANK        FL  721310    28-Feb-97   
...       ...                             ...       ...     ...          ...   
899133  44105                 NATL BK & TR CO        OH  322211    27-Feb-97   
899143   8053     BANCO POPULAR NORTH AMERICA        NJ  235610    27-Feb-97   
899151  18612  PNC BANK, NATIONAL ASSOCIATION        PA       0    27-Feb-97   
899158  75062   LOANS FROM OLD CLOSED LENDERS        DC       0    27-Feb-97   
899163  96734            CENTRAL PACIFIC BANK        HI       0    27-Feb-97   

       ApprovalFY  ...  RevLineCr  LowDoc  ChgOffDate  DisbursementDate  \
0            1997  ...          N       Y         NaN         28-Feb-99   
1            1997  ...          N       Y         NaN         31-May-97   
6            1980  ...          N       N   24-Jun-91         22-Jul-80   
7            1997  ...          N       Y         NaN         30-Jun-98   
8            1997  ...          N       N         NaN         31-Jul-97   
...           ...  ...        ...     ...         ...               ...   
899133       1997  ...          N       Y         NaN         31-May-97   
899143       1997  ...          0       N   19-Sep-00         31-May-97   
899151       1997  ...          Y       N         NaN         31-Oct-00   
899158       1997  ...          N       Y         NaN         30-Jun-97   
899163       1997  ...          N       N         NaN         31-May-97   

        DisbursementGross  BalanceGross  MIS_Status  ChgOffPrinGr  \
0             $60,000.00         $0.00        P I F        $0.00    
1             $40,000.00         $0.00        P I F        $0.00    
6            $600,000.00         $0.00       CHGOFF  $208,959.00    
7             $45,000.00         $0.00        P I F        $0.00    
8            $305,000.00         $0.00        P I F        $0.00    
...                   ...           ...         ...           ...   
899133        $99,000.00         $0.00        P I F        $0.00    
899143       $150,000.00         $0.00       CHGOFF  $124,847.00    
899151        $80,000.00         $0.00        P I F        $0.00    
899158        $79,000.00         $0.00        P I F        $0.00    
899163        $30,000.00         $0.00        P I F        $0.00    

              GrAppv      SBA_Appv  
0        $60,000.00    $48,000.00   
1        $40,000.00    $32,000.00   
6       $600,000.00   $499,998.00   
7        $45,000.00    $36,000.00   
8       $305,000.00   $228,750.00   
...              ...           ...  
899133   $99,000.00    $79,200.00   
899143  $150,000.00   $112,500.00   
899151   $80,000.00    $40,000.00  

In [21]:
df.query("NewExist != 0 and NewExist  != 1 and NewExist  != 2")

LoanNr_ChkDgt                            Name           City State  \
395231     3808405000               J.W.Chipping Corp      BETHLEHEM    NH   
399884     3854945010                D & R Foods Inc.      INDIANOLA    MS   
400052     3856405005         Luna Design Group, Inc.      LYNNFIELD    MA   
400203     3857955008               Lisbon Market Inc         LISBON    OH   
400300     3858915009         George F. Gales, Jr. MD         QUINCY    MA   
...               ...                             ...            ...   ...   
470308     4540125002                       GG&M, LLC        GENESEE    ID   
471939     4553585010            MORGAN VENTURES, INC        COLLINS    MS   
498244     4820745005                    Goodcare LLC    DELL RAPIDS    SD   
512594     4957935004  912 Enterprises Franchising, I  CRANBERRY TWP    PA   
611835     6080335003                  Michael McNabb    Hattieville    AR   

          Zip                       Bank BankState   NAICS ApprovalDate  \
395231   3574       FIRST COLEBROOK BANK        NH  423310    15-Dec-09   
399884  38751    TRUSTMARK NATIONAL BANK        MS  445110    13-Jan-10   
400052   1940       PEOPLE'S UNITED BANK        CT  541310    14-Jan-10   
400203  44432  THE CITIZENS SAVINGS BANK        OH  445110    15-Jan-10   
400300   2170               EASTERN BANK        MA  621111    15-Jan-10   
...       ...                        ...       ...     ...          ...   
470308  83832  ZIONS FIRST NATIONAL BANK        UT  424410    10-Feb-11   
471939  39428      COVINGTON COUNTY BANK        MS  722110    23-Feb-11   
498244  57022              MINNWEST BANK        SD  624310    12-Aug-11   
512594  16066    FIRST COMMONWEALTH BANK        PA  311520    16-Nov-11   
611835  72063       FIRST FINANCIAL BANK        AR  112320    29-Mar-13   

       ApprovalFY  ...  RevLineCr  LowDoc  ChgOffDate  DisbursementDate  \
395231       2010  ...          N       N         NaN         15-Dec-09   
399884       2010  ...          N       N         NaN          1-Feb-10   
400052       2010  ...          N       N         NaN         25-Mar-10   
400203       2010  ...          N       N         NaN         15-Jan-10   
400300       2010  ...          Y       N         NaN          1-Apr-10   
...           ...  ...        ...     ...         ...               ...   
470308       2011  ...          Y       N   28-Feb-13         10-Feb-11   
471939       2011  ...          N       N         NaN         23-Feb-11   
498244       2011  ...          T       N         NaN         19-Aug-11   
512594       2012  ...          N       N         NaN         16-Nov-11   
611835       2013  ...          N       N         NaN          1-Apr-13   

        DisbursementGross  BalanceGross  MIS_Status ChgOffPrinGr  \
395231       $117,000.00         $0.00        P I F       $0.00    
399884       $150,000.00         $0.00        P I F       $0.00    
400052        $25,000.00         $0.00        P I F       $0.00    
400203       $350,000.00         $0.00        P I F       $0.00    
400300        $52,395.00         $0.00        P I F       $0.00    
...                   ...           ...         ...          ...   
470308        $21,176.00         $0.00       CHGOFF  $12,842.00    
471939        $20,500.00         $0.00        P I F       $0.00    
498244       $129,717.00         $0.00        P I F       $0.00    
512594        $43,000.00         $0.00        P I F       $0.00    
611835        $50,700.00         $0.00        P I F       $0.00    

              GrAppv      SBA_Appv  
395231  $117,000.00    $99,450.00   
399884  $150,000.00   $135,000.00   
400052   $25,000.00    $25,000.00   
400203  $350,000.00   $175,000.00   
400300   $50,000.00    $25,000.00   
...              ...           ...  
470308   $15,000.00     $7,500.00   
471939   $20,500.00    $17,425.00   
498244  $100,000.00    $50,000.00   
512594   $43,000.00    $21,500.00   
611835   $50,700.00    $25,350.00   

[136 ro

In [ ]:
df["NewExist"].dtypes

In [23]:
df["NewExist"] = df["NewExist"].fillna(0) 

In [24]:
df["NewExist"].isna().any()

np.False_

In [ ]:
print(df.loc[df["NewExist"] == 1, "ApprovalDate"].sort_values(ascending=False))

In [ ]:
df["NewExist"].unique()

In [124]:
df["NewExist"] = df["NewExist"].replace(2, 1)

## Bank

In [ ]:
new_df["BankCount"] = new_df["Bank"].map(new_df["Bank"].value_counts())


In [ ]:
new_df[["small", "medium", "big"]]

In [ ]:
new_df["small"] = [1 if x < 1000 else 0 for x in new_df["BankCount"]]
new_df["medium"] = [1 if x > 1000 and x < 10000 else 0 for x in new_df["BankCount"]]
new_df["big"]= [1 if x > 10000 else 0 for x in new_df["BankCount"]]


In [ ]:
a = new_df[(new_df["BankState"].isnull()) & (new_df["State"] == "PA")]
a


In [ ]:
df[df["Bank"].isnull()]

In [ ]:
df[df["BankState"].isnull()]

## State

In [10]:
df.loc[(df["City"] == "PENNSVILLE") & (df["State"].isnull()), "State"] = "NY"
df.loc[(df["City"] == "JOHNSTOWN NY	") & (df["State"].isnull()), "State"] = "NY"
df.loc[(df["City"] == "CAMERON PARK") & (df["State"].isnull()), "State"] = "CA"
df.loc[(df["City"] == "WICHITA") & (df["State"].isnull()), "State"] = "KS"
df.loc[(df["City"] == "PALM BEACH GARDENS") & (df["State"].isnull()), "State"] = "FL"
df.loc[(df["City"] == "CASCO") & (df["State"].isnull()), "State"] = "WI"
df.loc[(df["City"] == "SOMERSET") & (df["State"].isnull()), "State"] = "WI"
df.loc[(df["City"] == "SALT LAKE CITY	") & (df["State"].isnull()), "State"] = "UT"
df.loc[(df["City"] == "LAKE OZARK	") & (df["State"].isnull()), "State"] = "MO"

In [11]:
df.loc[49244, "State"] = "NY"
df.loc[379174, "State"] = "UT"
df.loc[385418, "State"] = "MO"
df.loc[869948, "State"] = "TX"
df.loc[871847, "State"] = "TX"
df.loc[885335, "State"] = "TX"
df.loc[306274, "State"] = "HI"
df.loc[351072, "State"] = "TX"


#STATE OK --> PLUS DE NaN

In [ ]:
df["Zip"].isna().any()

In [ ]:
df.query("Zip == 0")

In [ ]:
df.query("Zip == 0")

In [ ]:
df.query("BankState != State").shape[0]

In [ ]:
df["Zip"].describe()

In [ ]:
a = df["State"].unique()
len(a)

In [ ]:
df["State"].unique()

In [402]:
df['north_east'] = [1 if x in ["MA", "NH", "NJ", "NY", "PA", "RI", "VT"] else 0 for x in df["State"]]
df['north_west'] = [1 if x in ["AK", "CO", "ID", "MT", "ND", "OR", "SD", "WA", "WY"] else 0 for x in df["State"]]
df['south_east'] = [1 if x in ["AL", "AR", "FL", "GA", "KY", "LA", "MD", "MS", "NC", "SC", "TN", "VA", "WV"] else 0 for x in df["State"]]
df['south_west'] = [1 if x in ["AZ", "CA", "CO", "HI", "NV", "NM", "OK", "TX"] else 0 for x in df["State"]]
df["midwest"] = [1 if x in ["IN", "IA", "IL", "KS", "MI", "MN", "MO", "NE", "ND", "OH", "SD", "WI"] else 0 for x in df["State"]]





In [ ]:
df.iloc[1]

In [ ]:
df["State"].unique()

In [ ]:
df[["north_east", "north_west", "south_east", "south_west", "midwest"]]
# df.drop("columns=State")

In [ ]:
df["Term"].isna().any()

In [ ]:
df["Term"].describe()

## term

In [35]:
a = df.iloc[:, -20:]


In [36]:

a.query("Term == 0")


NAICS ApprovalDate ApprovalFY  Term  NoEmp  NewExist  CreateJob  \
635     453220    25-Jun-01       2001     0      1       1.0          0   
673     424490     7-Feb-06       2006     0     10       1.0          7   
776     323110    10-Jan-03       2003     0     10       1.0          0   
916     315233    18-Apr-03       2003     0     20       1.0          0   
1084    532230    31-Jul-03       2003     0     11       1.0          0   
...        ...          ...        ...   ...    ...       ...        ...   
891986  812310     2-Feb-06       2006     0      2       1.0          2   
894913       0    16-Jan-97       1997     0      6       1.0          0   
895325  448310    22-Jan-97       1997     0     75       1.0          0   
896757       0     5-Feb-97       1997     0      5       2.0          0   
898765       0    25-Feb-97       1997     0      2       1.0          0   

        RetainedJob  FranchiseCode  UrbanRural RevLineCr LowDoc ChgOffDate  \
635               0              1           0         Y      N  24-Mar-06   
673              10              1           1         Y      N   6-Apr-11   
776               0              1           0         Y      N  23-Aug-10   
916               0              1           0         N      N  19-Sep-12   
1084              0              1           0         Y      N  27-Aug-10   
...             ...            ...         ...       ...    ...        ...   
891986            4              1           1         Y      N  24-Apr-14   
894913            0              1           0         N      Y   6-Jul-06   
895325            0              1           0         0      N   3-Aug-11   
896757            0              1           0         N      Y  15-Aug-02   
898765            0              1           0         Y      N   1-Jun-02   

       DisbursementDate DisbursementGross BalanceGross MIS_Status  \
635           30-Jun-01       $29,000.00        $0.00      CHGOFF   
673           31-Mar-06       $41,847.00        $0.00      CHGOFF   
776           31-Mar-03       $50,000.00        $0.00      CHGOFF   
916           30-Apr-03      $150,000.00        $0.00      CHGOFF   
1084          31-Aug-03      $432,428.00        $0.00      CHGOFF   
...                 ...               ...          ...        ...   
891986        30-Jun-06       $41,509.00        $0.00      CHGOFF   
894913        28-Feb-97       $93,000.00        $0.00      CHGOFF   
895325        31-Mar-97    $1,100,000.00        $0.00      CHGOFF   
896757        31-Jul-97       $40,000.00        $0.00      CHGOFF   
898765        28-Feb-01       $15,000.00        $0.00      CHGOFF   

        ChgOffPrinGr          GrAppv      SBA_Appv  
635       $2,416.00      $29,000.00    $14,500.00   
673      $19,695.00      $30,000.00    $15,000.00   
776      $15,546.00      $50,000.00    $25,000.00   
916      $33,634.00     $150,000.00    $75,000.00   
1084    $103,560.00     $150,000.00    $75,000.00   
...              ...             ...           ...  
891986    $6,646.00      $16,000.00     $8,000.00   
894913   $27,411.00      $93,000.00    $74,400.00   
895325  $124,314.00   $1,100,000.00   $748,000.00   
896757   $31,257.00      $40,000.00    $32,000.00   
898765   $13,686.00      $15,000.00     $7,500.00   

[810 rows x 20 columns]

In [37]:
df[df['Term'] == 0][["DisbursementGross","BalanceGross","MIS_Status"]]

DisbursementGross BalanceGross MIS_Status
635          $29,000.00        $0.00      CHGOFF
673          $41,847.00        $0.00      CHGOFF
776          $50,000.00        $0.00      CHGOFF
916         $150,000.00        $0.00      CHGOFF
1084        $432,428.00        $0.00      CHGOFF
...                  ...          ...        ...
891986       $41,509.00        $0.00      CHGOFF
894913       $93,000.00        $0.00      CHGOFF
895325    $1,100,000.00        $0.00      CHGOFF
896757       $40,000.00        $0.00      CHGOFF
898765       $15,000.00        $0.00      CHGOFF

[810 rows x 3 columns]

In [43]:
df[(df["Term"] == 0) & (df["GrAppv"] == 0)]

Empty DataFrame
Columns: [LoanNr_ChkDgt, Name, City, State, Zip, Bank, BankState, NAICS, ApprovalDate, ApprovalFY, Term, NoEmp, NewExist, CreateJob, RetainedJob, FranchiseCode, UrbanRural, RevLineCr, LowDoc, ChgOffDate, DisbursementDate, DisbursementGross, BalanceGross, MIS_Status, ChgOffPrinGr, GrAppv, SBA_Appv]
Index: []

[0 rows x 27 columns]

In [40]:
df[(df["Term"] == 0) & (df["MIS_Status"] == "PIF")][["DisbursementGross","BalanceGross"]]

Empty DataFrame
Columns: [DisbursementGross, BalanceGross]
Index: []

In [46]:
df[(df["Term"] == 0) & (df["MIS_Status"] != "PIF") & df["BalanceGross"] > 0][["BalanceGross"]]

BalanceGross
635          $0.00 
673          $0.00 
776          $0.00 
916          $0.00 
1084         $0.00 
...             ...
891986       $0.00 
894913       $0.00 
895325       $0.00 
896757       $0.00 
898765       $0.00 

[810 rows x 1 columns]

In [50]:
df[(df["Term"] == 0) & (df["MIS_Status"] != "PIF") & df["GrAppv"] > 0  ][["GrAppv"]]

GrAppv
635        $29,000.00 
673        $30,000.00 
776        $50,000.00 
916       $150,000.00 
1084      $150,000.00 
...                ...
891986     $16,000.00 
894913     $93,000.00 
895325  $1,100,000.00 
896757     $40,000.00 
898765     $15,000.00 

[810 rows x 1 columns]

In [44]:
df[(df["Term"] == 0)][["DisbursementDate"]]

DisbursementDate
635           30-Jun-01
673           31-Mar-06
776           31-Mar-03
916           30-Apr-03
1084          31-Aug-03
...                 ...
891986        30-Jun-06
894913        28-Feb-97
895325        31-Mar-97
896757        31-Jul-97
898765        28-Feb-01

[810 rows x 1 columns]

incoherence car 810 entreprises ont une date de décaissement, un montant décaissé etc mais une durée de prêt = 0 et surtout que des MIS_Status radié

 GrAppv > 0 → Un montant a bien été approuvé
✅ DisbursementDate renseignée → Le prêt a été versé
✅ BalanceGross = 0 → Plus aucun montant restant
✅ MIS_Status = CHGOFF → Prêt classé en défaut
✅ Term = 0 → Aucune durée définie

INterpretation : Prêts accordés et immédiatement en défaut 

    L'argent a été débloqué mais aucun remboursement n'a eu lieu.
    Peut être des prêts annulés ou des fraudes détectées après décaissement.

Erreur dans l'encodage de Term

    Un prêt ne peut pas avoir Term = 0 (durée nulle).
    Peut-être qu’un Term court (ex: 1-3 mois) a été mal enregistré.

Prêts "fantômes" ou annulations après décaissement

In [ ]:
df[(df["BalanceGross"] == 0) & (df["MIS_Status"] != "PIF") & (df["Term"] == 0) ][["DisbursementGross","BalanceGross"]]

In [39]:
df["Term"].describe()


count    899164.000000
mean        110.773078
std          78.857305
min           0.000000
25%          60.000000
50%          84.000000
75%         120.000000
max         569.000000
Name: Term, dtype: float64

In [ ]:
df[["Term", "BalanceGross"]]

In [ ]:
df["Term = 0"]= [1 if x == 0 else 0 for x in df["Term"]]


In [ ]:
df["Term = 0"].value_counts()

## noemp

In [32]:
df["NoEmp"].isna().any()

np.False_

In [11]:
df["NoEmp"].describe()

count    899164.000000
mean         11.411353
std          74.108196
min           0.000000
25%           2.000000
50%           4.000000
75%          10.000000
max        9999.000000
Name: NoEmp, dtype: float64

In [ ]:
a = df.iloc[:,-17:]

In [33]:
df.query("NoEmp == 0")

LoanNr_ChkDgt                            Name          City State  \
1089       1007485003   Hoffman and Associates Dental       ORLANDO    FL   
1092       1007495006   Hoffman and Associates Dental       ORLANDO    FL   
1463       1009885009             THE TIRE DEPOT, INC       BRANDON    MS   
1517       1010355000  Tropical Smoothie Cafe and Del       MADISON    AL   
1590       1010865000  The Matterhorn - Ice Cream & C  COLLEGEVILLE    PA   
...               ...                             ...           ...   ...   
664272     6728795005                    JH Group LLC    PEARL CITY    HI   
669507     6816425010         Toasted Frog Fargo, LLC         FARGO    ND   
672918     6878215001         Onsite Contractors, LLC      Harrison    OH   
673911     6895265007      MARK DEMATTIA INCORPORATED     EWA BEACH    HI   
726149     7556880003                 RALPH O BURNETT    BAKERSTOWN    PA   

          Zip                            Bank BankState   NAICS ApprovalDate  \
1089    32826     WELLS FARGO BANK NATL ASSOC        SD  339116     1-Aug-03   
1092    32826     WELLS FARGO BANK NATL ASSOC        SD  339116     1-Aug-03   
1463    39047         TRUSTMARK NATIONAL BANK        MS  441320    15-Jan-04   
1517    35758     WELLS FARGO BANK NATL ASSOC        SD  722211     9-Mar-04   
1590    19426      UNIVEST BANK AND TRUST CO.        PA  722213    26-Mar-04   
...       ...                             ...       ...     ...          ...   
664272  96782      AMERICAN SAVINGS BANK, FSB        HI  722511    18-Mar-14   
669507  58105                    BANK FORWARD        ND  722511     6-May-14   
672918  45030    THE HUNTINGTON NATIONAL BANK        OH  237110     5-Jun-14   
673911  96706            CENTRAL PACIFIC BANK        HI  492210    12-Jun-14   
726149  15007  PNC BANK, NATIONAL ASSOCIATION        PA       0    28-Apr-69   

       ApprovalFY  ...  RevLineCr  LowDoc  ChgOffDate  DisbursementDate  \
1089         2003  ...          N       N         NaN         30-Sep-03   
1092         2003  ...          Y       N         NaN         31-Dec-03   
1463         2004  ...          Y       N         NaN         31-Aug-04   
1517         2004  ...          N       N         NaN         31-Aug-04   
1590         2004  ...          Y       N         NaN         31-Jul-04   
...           ...  ...        ...     ...         ...               ...   
664272       2014  ...          N       N         NaN         18-Mar-14   
669507       2014  ...          N       N         NaN         30-Jun-14   
672918       2014  ...          Y       N         NaN          5-Jun-14   
673911       2014  ...          N       N         NaN         12-Jun-14   
726149       1969  ...          N       N   21-Jun-95         22-May-69   

        DisbursementGross  BalanceGross  MIS_Status ChgOffPrinGr  \
1089          $86,000.00         $0.00        P I F       $0.00    
1092          $36,857.00         $0.00        P I F       $0.00    
1463          $84,168.00         $0.00        P I F       $0.00    
1517         $125,000.00         $0.00        P I F       $0.00    
1590          $65,003.00         $0.00        P I F       $0.00    
...                   ...           ...         ...          ...   
664272       $200,000.00         $0.00        P I F       $0.00    
669507       $150,000.00         $0.00        P I F       $0.00    
672918        $17,200.00         $0.00        P I F       $0.00    
673911        $50,000.00         $0.00        P I F       $0.00    
726149        $27,200.00         $0.00        P I F       $0.00    

              GrAppv      SBA_Appv  
1089     $86,000.00    $43,000.00   
1092     $20,000.00    $10,000.00   
1463     $50,000.00    $25,000.00   
1517    $125,000.00    $62,500.00   
1590     $45,000.00    $22,500.00   
...              ...           ...  
664272  $200,000.00   $180,000.00   
669507  $150,000.00    $75,000.00   
672918   $17,200.00     $8,600.00   
673911   $50,000.00    $25,000.00 

In [12]:
df[(df["NoEmp"] == 0) & (df["CreateJob"] != 0)]

LoanNr_ChkDgt                            Name          City State  \
15292      1110875000         JD's Subway 35210, Inc.    LOUISVILLE    KY   
20038      1147125002  SANDY'S BACK PORCH GARDEN CENT    BELLEVILLE    IL   
20536      1151895002  Sleep Insights Medical Service      FAIRPORT    NY   
20762      1153745007                   Thomas Jaindl  NORTHBOROUGH    MA   
21390      1158595009       Pacific Coast Market, LLC  Indianapolis    IN   
...               ...                             ...           ...   ...   
650892     6530715008           The Paint Factory LLC    HUTCHINSON    MN   
654244     6581605010  South Hill Food Inc. dba Wetze        Humble    TX   
657662     6633895008           No Coast Reserve, Inc   BLOOMINGTON    IN   
664272     6728795005                    JH Group LLC    PEARL CITY    HI   
669507     6816425010         Toasted Frog Fargo, LLC         FARGO    ND   

          Zip                            Bank BankState   NAICS ApprovalDate  \
15292   40202     WELLS FARGO BANK NATL ASSOC        SD  722211    22-Nov-04   
20038   62226      ASSOCIATED BANK NATL ASSOC        WI  444220     3-Jan-05   
20536   14450   MANUFACTURERS & TRADERS TR CO        NY  621999     7-Jan-05   
20762    1532    FLAGSHIP BANK, A DIVISION OF        MA  451110    10-Jan-05   
21390   46254  PNC BANK, NATIONAL ASSOCIATION        OH  424490    13-Jan-05   
...       ...                             ...       ...     ...          ...   
650892  55350        CITIZENS BANK & TRUST CO        MN  711510    13-Nov-13   
654244  77338               PACIFIC CITY BANK        CA  722513    11-Dec-13   
657662  47408               OLD NATIONAL BANK        IN  722511    15-Jan-14   
664272  96782      AMERICAN SAVINGS BANK, FSB        HI  722511    18-Mar-14   
669507  58105                    BANK FORWARD        ND  722511     6-May-14   

       ApprovalFY  ...  RevLineCr  LowDoc  ChgOffDate  DisbursementDate  \
15292        2005  ...          N       N         NaN         31-Dec-04   
20038        2005  ...          N       N         NaN         28-Feb-05   
20536        2005  ...          N       N         NaN         31-May-05   
20762        2005  ...          N       N         NaN         31-Mar-05   
21390        2005  ...          Y       N         NaN         31-Jan-05   
...           ...  ...        ...     ...         ...               ...   
650892       2014  ...          N       N         NaN         13-Nov-13   
654244       2014  ...          N       N         NaN          1-Jan-14   
657662       2014  ...          N       N         NaN          1-Mar-14   
664272       2014  ...          N       N         NaN         18-Mar-14   
669507       2014  ...          N       N         NaN         30-Jun-14   

        DisbursementGross  BalanceGross  MIS_Status ChgOffPrinGr  \
15292        $135,000.00         $0.00        P I F       $0.00    
20038         $64,000.00         $0.00        P I F       $0.00    
20536        $250,000.00         $0.00        P I F       $0.00    
20762        $340,000.00         $0.00        P I F       $0.00    
21390         $49,236.00         $0.00        P I F       $0.00    
...                   ...           ...         ...          ...   
650892        $22,000.00         $0.00        P I F       $0.00    
654244       $115,000.00         $0.00        P I F       $0.00    
657662       $195,000.00         $0.00        P I F       $0.00    
664272       $200,000.00         $0.00        P I F       $0.00    
669507       $150,000.00         $0.00        P I F       $0.00    

              GrAppv      SBA_Appv  
15292   $135,000.00    $67,500.00   
20038    $64,000.00    $32,000.00   
20536   $250,000.00   $187,500.00   
20762   $340,000.00   $255,000.00   
21390    $40,000.00    $20,000.00   
...              ...           ...  
650892   $22,000.00    $11,000.00   
654244  $115,000.00    $97,750.00   
657662  $195,000.00   $146,250.00   
664272  $200,000.00   $180,000.00 

## createjob

In [121]:
df["CreateJob"].isna().any()

np.False_

In [3]:
df['CreateJob'].describe()

count    899164.000000
mean          8.430376
std         236.688165
min           0.000000
25%           0.000000
50%           0.000000
75%           1.000000
max        8800.000000
Name: CreateJob, dtype: float64

In [124]:
df["CreateJob"].value_counts()

CreateJob
0       629248
1        63174
2        57831
3        28806
4        20511
         ...  
2515         1
162          1
182          1
1016         1
860          1
Name: count, Length: 246, dtype: int64

In [103]:
a = df.iloc[:,-15:]

In [ ]:
df[(df["CreateJob"] > 1000)]

In [140]:
df[(df["CreateJob"] < 10)]

LoanNr_ChkDgt                            Name             City State  \
0          1000014003                  ABC HOBBYCRAFT       EVANSVILLE    IN   
1          1000024006     LANDMARK BAR & GRILLE (THE)        NEW PARIS    IN   
2          1000034009           WHITLOCK DDS, TODD M.      BLOOMINGTON    IN   
3          1000044001   BIG BUCKS PAWN & JEWELRY, LLC     BROKEN ARROW    OK   
4          1000054004     ANASTASIA CONFECTIONS, INC.          ORLANDO    FL   
...               ...                             ...              ...   ...   
899159     9995573004                    FABRIC FARMS  UPPER ARLINGTON    OH   
899160     9995603000                    FABRIC FARMS         COLUMBUS    OH   
899161     9995613003    RADCO MANUFACTURING CO.,INC.      SANTA MARIA    CA   
899162     9995973006           MARUTAMA HAWAII, INC.         HONOLULU    HI   
899163     9996003010  PACIFIC TRADEWINDS FAN & LIGHT           KAILUA    HI   

          Zip                            Bank BankState   NAICS ApprovalDate  \
0       47711                FIFTH THIRD BANK        OH  451120    28-Feb-97   
1       46526                 1ST SOURCE BANK        IN  722410    28-Feb-97   
2       47401         GRANT COUNTY STATE BANK        IN  621210    28-Feb-97   
3       74012   1ST NATL BK & TR CO OF BROKEN        OK       0    28-Feb-97   
4       32801         FLORIDA BUS. DEVEL CORP        FL       0    28-Feb-97   
...       ...                             ...       ...     ...          ...   
899159  43221  JPMORGAN CHASE BANK NATL ASSOC        IL  451120    27-Feb-97   
899160  43221  JPMORGAN CHASE BANK NATL ASSOC        IL  451130    27-Feb-97   
899161  93455  RABOBANK, NATIONAL ASSOCIATION        CA  332321    27-Feb-97   
899162  96830                  BANK OF HAWAII        HI       0    27-Feb-97   
899163  96734            CENTRAL PACIFIC BANK        HI       0    27-Feb-97   

       ApprovalFY  ...  RevLineCr  LowDoc  ChgOffDate  DisbursementDate  \
0            1997  ...          N       Y         NaN         28-Feb-99   
1            1997  ...          N       Y         NaN         31-May-97   
2            1997  ...          N       N         NaN         31-Dec-97   
3            1997  ...          N       Y         NaN         30-Jun-97   
4            1997  ...          N       N         NaN         14-May-97   
...           ...  ...        ...     ...         ...               ...   
899159       1997  ...          0       N         NaN         30-Sep-97   
899160       1997  ...          Y       N         NaN         31-Oct-97   
899161       1997  ...          N       N         NaN         30-Sep-97   
899162       1997  ...          N       Y    8-Mar-00         31-Mar-97   
899163       1997  ...          N       N         NaN         31-May-97   

        DisbursementGross  BalanceGross  MIS_Status ChgOffPrinGr  \
0             $60,000.00         $0.00        P I F       $0.00    
1             $40,000.00         $0.00        P I F       $0.00    
2            $287,000.00         $0.00        P I F       $0.00    
3             $35,000.00         $0.00        P I F       $0.00    
4            $229,000.00         $0.00        P I F       $0.00    
...                   ...           ...         ...          ...   
899159        $70,000.00         $0.00        P I F       $0.00    
899160        $85,000.00         $0.00        P I F       $0.00    
899161       $300,000.00         $0.00        P I F       $0.00    
899162        $75,000.00         $0.00       CHGOFF  $46,383.00    
899163        $30,000.00         $0.00        P I F       $0.00    

              GrAppv      SBA_Appv  
0        $60,000.00    $48,000.00   
1        $40,000.00    $32,000.00   
2       $287,000.00   $215,250.00   
3        $35,000.00    $28,000.00   
4       $229,000.00   $229,000.00   
...              ...           ...  
899159   $70,000.00    $56,000.00   
899160   $85,000.00    $42,500.00   
899161  $300,000.00   $225,000.00  

In [133]:
lower_percentile = df["CreateJob"].quantile(0.25)
upper_percentile = df["CreateJob"].quantile(0.75)
extreme_threshold = df["CreateJob"].quantile(0.95)


In [141]:
print(lower_percentile, upper_percentile,extreme_threshold)


0.0 1.0 10.0


In [128]:
df[(df["CreateJob"] == 8800) & (df["NewExist"] == 2)]

LoanNr_ChkDgt                            Name              City State  \
235160     2662583007   JOHNNYCAKE HILL ASSOCIATES, A       MIDDLETOWN,    RI   
235416     2664373007     MCKAY LINCOLN MERCURY, INC.       COON RAPIDS    MN   
235423     2664403003                FOUR CROWN, INC.  OAK PARK HEIGHTS    MN   
235514     2665093008              STAR TROPICS, INC.     HICKORY HILLS    IL   
235560     2665373009            AUTOBRITE - BROADWAY             BOISE    ID   
...               ...                             ...               ...   ...   
244819     2726403010  DAYTON PAINT AND COATINGS, INC            DAYTON    OH   
244874     2726793000          SUNSHINE DENTAL CENTER  COLORADO SPRINGS    CO   
244925     2727183007  PILGREENS WEST & SISSON ENT IN      DOUGLASVILLE    GA   
244936     2727273010                  AJ'S DELI MART            YAKIMA    WA   
244948     2727363002             BUTCHER BOY GROCERY           PALMYRA    IN   

          Zip                            Bank BankState  NAICS ApprovalDate  \
235160   2840  OCEAN ST. BUS. DEVEL AUTHORITY        RI      0    30-Jun-86   
235416  55433      CENTRAL MINNESOTA DEVEL CO        MN      0     2-Jul-86   
235423  55082       MINNESOTA BUS. FINAN CORP        MN      0     2-Jul-86   
235514  60465            NORTH COMMUNITY BANK        IL      0     2-Jul-86   
235560  83707            CAPITAL MATRIX, INC.        ID      0     2-Jul-86   
...       ...                             ...       ...    ...          ...   
244819  45403  CITYWIDE SMALL BUS. DEVEL CORP        OH      0    29-Sep-86   
244874  80904  PIKES PEAK REGIONAL DEVEL CORP        CO      0    30-Sep-86   
244925  30135             SB&T, A DIVISION OF        GA      0    30-Sep-86   
244936  98907      EVERGREEN BUSINESS CAPITAL        WA      0    30-Sep-86   
244948  47164  PNC BANK, NATIONAL ASSOCIATION        DE      0    30-Sep-86   

       ApprovalFY  ...  RevLineCr  LowDoc  ChgOffDate  DisbursementDate  \
235160       1986  ...          N       N         NaN         10-Dec-86   
235416       1986  ...          N       N         NaN         17-Jun-87   
235423       1986  ...          N       N         NaN         12-Nov-86   
235514       1986  ...          N       N   20-Apr-90          7-Oct-86   
235560       1986  ...          N       N         NaN         15-Apr-87   
...           ...  ...        ...     ...         ...               ...   
244819       1986  ...          N       N         NaN         15-Jul-87   
244874       1986  ...          N       N         NaN         11-Feb-87   
244925       1986  ...          N       N    1-Nov-90         22-Oct-86   
244936       1986  ...          N       N         NaN         13-May-87   
244948       1986  ...          N       N   26-Jul-91         16-Oct-86   

        DisbursementGross  BalanceGross  MIS_Status  ChgOffPrinGr  \
235160       $227,000.00         $0.00        P I F        $0.00    
235416       $486,000.00         $0.00        P I F        $0.00    
235423       $233,000.00         $0.00        P I F        $0.00    
235514        $65,000.00         $0.00       CHGOFF   $62,735.00    
235560       $358,000.00         $0.00        P I F        $0.00    
...                   ...           ...         ...           ...   
244819        $88,000.00         $0.00        P I F        $0.00    
244874       $109,000.00         $0.00        P I F        $0.00    
244925       $525,000.00         $0.00       CHGOFF  $180,374.00    
244936       $180,000.00         $0.00        P I F        $0.00    
244948       $138,000.00         $0.00       CHGOFF   $90,465.00    

              GrAppv      SBA_Appv  
235160  $227,000.00   $227,000.00   
235416  $500,000.00   $500,000.00   
235423  $233,000.00   $233,000.00   
235514   $65,000.00    $57,525.00   
235560  $358,000.00   $358,000.00   
...              ...           ...  
244819   $88,000.00    $88,000.00   
244874  $109,000.00   $109,000.00   
244925  $525,000.00   $

In [ ]:
a.query("CreateJob == 8800")


In [147]:
df["FranchiseCode"].unique()

array([    1,     0, 15100, ...,  2899, 18701, 15930], shape=(2768,))

In [ ]:
a = df.iloc[:,-13:]
a

In [ ]:
a.query("FranchiseCode > 2")

In [14]:
df["FranchiseCode"].isnull().any()

np.False_

In [ ]:
df[df["FranchiseCode"] > 2]

In [ ]:
df[(df["FranchiseCode"] > 2 ) & (df["CreateJob"] > 7000) & (df["UrbanRural"] == 0 )]

#entreprise qui a créer plus de 7000 emplois, qui est une franchise et qui est en zone urbaine (resultat =0 si zone rurale)
#33 entreprises

In [ ]:
df[(df["FranchiseCode"] < 2 ) & (df["CreateJob"] > 8000) & (df["UrbanRural"] == 0 )]

#entreprise qui a créer plus de 7000 emplois, qui n'est pas une franchise et qui est en zone urbaine 
#resultat 615

In [125]:
df[df["CreateJob"] > 7000]["DisbursementGross"].describe()


count              648
unique             313
top       $500,000.00 
freq                31
Name: DisbursementGross, dtype: object

In [ ]:
df[df["CreateJob"] < 7000]["DisbursementGross"].describe()


In [ ]:
df[df["CreateJob"] > 7000][["CreateJob", "RetainedJob"]].describe()


In [ ]:
df[df["CreateJob"] > 7000]["State"].value_counts()


## RetainedJob

In [ ]:
df["RetainedJob"].describe()

In [191]:
a = df.iloc[:,-15:]

In [ ]:
a.head()

In [ ]:
a[a["RetainedJob"] > 7000 ]

In [ ]:
df[df["CreateJob"] > 7000]

In [202]:
df1 = df.iloc[:,-15:]

In [ ]:
df1[
(df1["CreateJob"] == 8800) & (df1["RetainedJob"] == 8800)
].sort_values(by="GrAppv", ascending=False)

## urbanrural

In [148]:
df["UrbanRural"].isna().any()

np.False_

In [149]:
df["UrbanRural"].describe()

count    899164.000000
mean          0.757748
std           0.646436
min           0.000000
25%           0.000000
50%           1.000000
75%           1.000000
max           2.000000
Name: UrbanRural, dtype: float64

In [ ]:
df[(df["UrbanRural"]==0) & (df["State"] == "CA")
   
   
   
   ]

In [142]:
df[(df["UrbanRural"]==0)]

LoanNr_ChkDgt                            Name             City State  \
0          1000014003                  ABC HOBBYCRAFT       EVANSVILLE    IN   
1          1000024006     LANDMARK BAR & GRILLE (THE)        NEW PARIS    IN   
2          1000034009           WHITLOCK DDS, TODD M.      BLOOMINGTON    IN   
3          1000044001   BIG BUCKS PAWN & JEWELRY, LLC     BROKEN ARROW    OK   
4          1000054004     ANASTASIA CONFECTIONS, INC.          ORLANDO    FL   
...               ...                             ...              ...   ...   
899159     9995573004                    FABRIC FARMS  UPPER ARLINGTON    OH   
899160     9995603000                    FABRIC FARMS         COLUMBUS    OH   
899161     9995613003    RADCO MANUFACTURING CO.,INC.      SANTA MARIA    CA   
899162     9995973006           MARUTAMA HAWAII, INC.         HONOLULU    HI   
899163     9996003010  PACIFIC TRADEWINDS FAN & LIGHT           KAILUA    HI   

          Zip                            Bank BankState   NAICS ApprovalDate  \
0       47711                FIFTH THIRD BANK        OH  451120    28-Feb-97   
1       46526                 1ST SOURCE BANK        IN  722410    28-Feb-97   
2       47401         GRANT COUNTY STATE BANK        IN  621210    28-Feb-97   
3       74012   1ST NATL BK & TR CO OF BROKEN        OK       0    28-Feb-97   
4       32801         FLORIDA BUS. DEVEL CORP        FL       0    28-Feb-97   
...       ...                             ...       ...     ...          ...   
899159  43221  JPMORGAN CHASE BANK NATL ASSOC        IL  451120    27-Feb-97   
899160  43221  JPMORGAN CHASE BANK NATL ASSOC        IL  451130    27-Feb-97   
899161  93455  RABOBANK, NATIONAL ASSOCIATION        CA  332321    27-Feb-97   
899162  96830                  BANK OF HAWAII        HI       0    27-Feb-97   
899163  96734            CENTRAL PACIFIC BANK        HI       0    27-Feb-97   

       ApprovalFY  ...  RevLineCr  LowDoc  ChgOffDate  DisbursementDate  \
0            1997  ...          N       Y         NaN         28-Feb-99   
1            1997  ...          N       Y         NaN         31-May-97   
2            1997  ...          N       N         NaN         31-Dec-97   
3            1997  ...          N       Y         NaN         30-Jun-97   
4            1997  ...          N       N         NaN         14-May-97   
...           ...  ...        ...     ...         ...               ...   
899159       1997  ...          0       N         NaN         30-Sep-97   
899160       1997  ...          Y       N         NaN         31-Oct-97   
899161       1997  ...          N       N         NaN         30-Sep-97   
899162       1997  ...          N       Y    8-Mar-00         31-Mar-97   
899163       1997  ...          N       N         NaN         31-May-97   

        DisbursementGross  BalanceGross  MIS_Status ChgOffPrinGr  \
0             $60,000.00         $0.00        P I F       $0.00    
1             $40,000.00         $0.00        P I F       $0.00    
2            $287,000.00         $0.00        P I F       $0.00    
3             $35,000.00         $0.00        P I F       $0.00    
4            $229,000.00         $0.00        P I F       $0.00    
...                   ...           ...         ...          ...   
899159        $70,000.00         $0.00        P I F       $0.00    
899160        $85,000.00         $0.00        P I F       $0.00    
899161       $300,000.00         $0.00        P I F       $0.00    
899162        $75,000.00         $0.00       CHGOFF  $46,383.00    
899163        $30,000.00         $0.00        P I F       $0.00    

              GrAppv      SBA_Appv  
0        $60,000.00    $48,000.00   
1        $40,000.00    $32,000.00   
2       $287,000.00   $215,250.00   
3        $35,000.00    $28,000.00   
4       $229,000.00   $229,000.00   
...              ...           ...  
899159   $70,000.00    $56,000.00   
899160   $85,000.00    $42,500.00   
899161  $300,000.00   $225,000.00  

In [143]:
df[(df["UrbanRural"]==1)]

LoanNr_ChkDgt                            Name         City State  \
12         1000146010                          CARVEL         APEX    NC   
24         1000446006                          SUBWAY  LITTLE ROCK    AR   
28         1000506000             WEYLAND CORPORATION    CAMARILLO    CA   
30         1000516003     CHICAGO BRICK UNLIMITED INC        MIAMI    FL   
32         1000526006                       RZI, INC.  NEW ORLEANS    LA   
...               ...                             ...          ...   ...   
896389     9933753004                   HEADS & TAILS      ANTIOCH    CA   
898240     9976073001      CAFFE' DIVA ESPRESSO, INC.       EUGENE    OR   
898660     9984863002  SISTERS' & BROTHERS' BOOKSTORE  ALBUQUERQUE    NM   
898731     9986753008                         ISSALAK        SALEM    MA   
898998     9992733009                   DESIGNER DOGS       OAKLAY    CA   

          Zip                            Bank BankState   NAICS ApprovalDate  \
12      27502           STEARNS BK NATL ASSOC        MN  445299     7-Feb-06   
24      72223                        HOPE FCU        MS  722211     7-Feb-06   
28      93010     WELLS FARGO BANK NATL ASSOC        SD  611110     7-Feb-06   
30      33186                  CITIBANK, N.A.        FL  238140     7-Feb-06   
32      70130  BUSINESS RES. CAP. SPECIALTY B        LA  532490     7-Feb-06   
...       ...                             ...       ...     ...          ...   
896389  94509  BANK OF AGRICULTURE & COMMERCE        CA       0    30-Jan-97   
898240  97401        PACIFIC CONTINENTAL BANK        OR       0    20-Feb-97   
898660  87108     WELLS FARGO BANK NATL ASSOC        SD  451211    24-Feb-97   
898731   1970     NATL GRAND BK OF MARBLEHEAD        MA       0    25-Feb-97   
898998  94561                 INNOVATIVE BANK        CA       0    27-Feb-97   

       ApprovalFY  ...  RevLineCr  LowDoc  ChgOffDate  DisbursementDate  \
12           2006  ...          N       N         NaN         31-Mar-06   
24           2006  ...          N       N         NaN         30-Apr-06   
28           2006  ...          Y       N         NaN         28-Feb-06   
30           2006  ...          Y       N         NaN         28-Feb-06   
32           2006  ...          N       N         NaN         31-May-06   
...           ...  ...        ...     ...         ...               ...   
896389       1997  ...          N       N         NaN         28-Feb-97   
898240       1997  ...          N       N   26-Jul-02         28-Feb-97   
898660       1997  ...          N       N         NaN         31-Mar-97   
898731       1997  ...          N       N         NaN         30-Apr-97   
898998       1997  ...          N       N         NaN         30-Sep-97   

        DisbursementGross  BalanceGross  MIS_Status ChgOffPrinGr  \
12           $253,400.00         $0.00        P I F       $0.00    
24           $137,300.00         $0.00        P I F       $0.00    
28           $438,541.00         $0.00        P I F       $0.00    
30            $51,440.00         $0.00        P I F       $0.00    
32            $50,000.00         $0.00        P I F       $0.00    
...                   ...           ...         ...          ...   
896389        $12,218.00         $0.00        P I F       $0.00    
898240       $100,000.00         $0.00       CHGOFF   $9,771.00    
898660        $57,000.00         $0.00        P I F       $0.00    
898731        $45,000.00         $0.00        P I F       $0.00    
898998        $86,000.00         $0.00        P I F       $0.00    

              GrAppv      SBA_Appv  
12      $253,400.00   $190,050.00   
24      $137,300.00   $116,705.00   
28      $100,000.00    $50,000.00   
30       $35,000.00    $17,500.00   
32       $50,000.00    $42,500.00   
...              ...           ...  
896389   $20,000.00    $16,000.00   
898240  $100,000.00    $80,000.00   
898660   $57,000.00    $45,600.00   
898731   $45,000.00    $36,000.00   
898998   

In [145]:
df["FranchiseCode"].isna().any()

np.False_

In [4]:
df['UrbanRural'].unique()

array([0, 1, 2])

In [144]:
df[(df["UrbanRural"]==2)]

LoanNr_ChkDgt                            Name  \
34         1000536009          HUTMACHER HOLDINGS INC   
67         1000766008                LILY DAY GARDENS   
76         1000815003  Robert's Realty of Victor Inc.   
89         1000896001                  SHAFI SECURITY   
127        1001136009   SPOHN, INC DBA CULVER'S OF MT   
...               ...                             ...   
894747     9897044001              IVEY'S MOTOR LODGE   
894766     9897464009                BACK YARD DREAMS   
894798     9898024008            SPRAYKING CONTRACTOR   
894801     9898054006               5280 MARTIAL ARTS   
896082     9927983007        TUBS TO GO SPA MFG., LLP   

                                  City State    Zip  \
34                             LEANDER    TX  78641   
67                              BURNET    TX  78611   
76                              Victor    MT  59875   
89                            THORNTON    CO  80602   
127                        MT PLEASANT    MI  48858   
...                                ...   ...    ...   
894747  HOULTON (CENSUS NAME FOR HOULT    ME   4730   
894766                         ROCKLIN    CA  95765   
894798                       OCEANSIDE    CA  92057   
894801                 HIGHLANDS RANCH    CO  80230   
896082                          HELENA    MT  59635   

                                  Bank BankState   NAICS ApprovalDate  \
34         WELLS FARGO BANK NATL ASSOC        SD  541611     7-Feb-06   
67         WELLS FARGO BANK NATL ASSOC        SD  444220     7-Feb-06   
76         WELLS FARGO BANK NATL ASSOC        SD       0     3-Apr-97   
89      JPMORGAN CHASE BANK NATL ASSOC        IL  561612     7-Feb-06   
127             WISCONSIN BANK & TRUST        WI  722211     7-Feb-06   
...                                ...       ...     ...          ...   
894747        EASTERN MAINE DEVEL CORP        ME  721110     6-Feb-06   
894766       CDC SMALL BUS. FINAN CORP        CA  332611     6-Feb-06   
894798  JPMORGAN CHASE BANK NATL ASSOC        IL  236220     6-Feb-06   
894801  JPMORGAN CHASE BANK NATL ASSOC        IL  611620     6-Feb-06   
896082  VALLEY BK OF HELENA A DIVISION        MT  326191    29-Jan-97   

       ApprovalFY  ...  RevLineCr  LowDoc  ChgOffDate  DisbursementDate  \
34           2006  ...          Y       N         NaN         31-May-06   
67           2006  ...          Y       N   27-Jan-11         28-Feb-06   
76           1997  ...          0       N         NaN         31-Aug-97   
89           2006  ...          Y       N   26-Sep-08         28-Feb-06   
127          2006  ...          N       N         NaN         28-Feb-06   
...           ...  ...        ...     ...         ...               ...   
894747       2006  ...          0       N         NaN         11-Oct-06   
894766       2006  ...          0       N         NaN         11-Jul-07   
894798       2006  ...          Y       N         NaN         30-Apr-06   
894801       2006  ...          Y       N         NaN         28-Feb-06   
896082       1997  ...          N       N    1-Sep-98         31-Mar-97   

        DisbursementGross  BalanceGross  MIS_Status  ChgOffPrinGr  \
34            $63,076.00         $0.00        P I F        $0.00    
67            $60,859.00         $0.00       CHGOFF   $19,994.00    
76            $25,000.00         $0.00        P I F        $0.00    
89            $76,432.00         $0.00       CHGOFF   $28,979.00    
127          $150,000.00         $0.00        P I F        $0.00    
...                   ...           ...         ...           ...   
894747       $347,000.00         $0.00        P I F        $0.00    
894766       $134,000.00         $0.00        P I F        $0.00    
894798        $42,399.00         $0.00        P I F        $0.00    
894801        $30,000.00         $0.00        P I F        $0.00    
896082       $170,000.00         $0.00       CHGOFF  $170,124.00    

              GrAppv      SBA_Appv  
34       $25,000.0

## ChgOffDate

In [ ]:
df["ChgOffDate"].isna().sum()

In [ ]:
df.loc[14000,'ChgOffDate']

In [ ]:
df["ChgOffDate"] = pd.to_datetime(df["ChgOffDate"])


In [275]:
df["ChgOffDate"] = df["ChgOffDate"].dt.to_period("M")


In [ ]:
df[["ChgOffDate"]].head(10)

In [ ]:
df["ChgOffDate"].notnull().sum()
#date présente donc radié


In [ ]:
df[(df["MIS_Status"] == "CHGOFF")]

In [ ]:
df["MIS_Status"].unique()

In [ ]:
df['ChgOffDate'].describe()

## DisbursementDate

In [ ]:
df["DisbursementDate"].isna().any()

In [ ]:
df["DisbursementDate"].isna().sum()

In [ ]:
df[(df["DisbursementDate"].isna())]

In [ ]:
df.loc[452]

In [ ]:
df.loc[611]

In [ ]:
df[["DisbursementDate","ApprovalDate", "ApprovalFY"]].head(25)

In [10]:
df["DisbursementDate"] = df["DisbursementDate"].fillna(df["ApprovalDate"]) 


## ApprovalDate

In [ ]:
df["ApprovalDate"] = pd.to_datetime(df["ApprovalDate"])

In [280]:
df["ApprovalDate"] = df["ApprovalDate"].dt.to_period("M")

In [ ]:
df["ApprovalDate"].head(10)

## DisbursementGross

In [ ]:
df["DisbursementGross"].isna().any()

In [ ]:
df["DisbursementGross"].describe()

In [ ]:
df.loc[df["NewExist"] == 1, "ApprovalDate"].sort_values(ascending=False)

In [ ]:
df["DisbursementGross"].sort_values(ascending=False)

In [26]:
df["DisbursementGross"] = df["DisbursementGross"].replace({"$": "", ",": ""}, regex=True)

In [ ]:
df.loc[21358,"DisbursementGross"]

In [ ]:
df[

(df["DisbursementGross"] == "$1.00 ")
]

## LowDoc

In [155]:
df["LowDoc"].isna().any()

np.True_

In [160]:
df["LowDoc"].unique()

array(['Y', 'N'], dtype=object)

(0, 1, A, C, R, S); 
Y = Y, 1
N = 
moins de 150k 1 page donc Y


In [ ]:
df['DisbursementGross'].isna().any()

In [ ]:
df[["DisbursementGross", "GrAppv"]]

In [158]:
df["DisbursementGross"] = df["DisbursementGross"].str.replace("$", "").str.replace(",", "").str.replace(" ","").astype(float)

In [159]:
df["LowDoc"] = ["Y" if x < 150000 else "N" for x in df["DisbursementGross"]]


## GrAppv

In [161]:
df["GrAppv"].isna().any()

np.False_

In [164]:
df["GrAppv"].head(7)

0     60000.0
1     40000.0
2    287000.0
3     35000.0
4    229000.0
5    517000.0
6    600000.0
Name: GrAppv, dtype: float64

In [163]:
df["GrAppv"] = df["GrAppv"].str.replace("$", "").str.replace(",", "").str.replace(" ","").astype(float)


## revline

In [154]:
df['RevLineCr'].isnull().any()

np.False_

In [153]:
df['RevLineCr'].unique()

array(['N', 'Y'], dtype=object)

In [152]:
df["RevLineCr"] = ["Y" if x == "Y" or x == "1" else "N" for x in df["RevLineCr"]]

In [18]:
df[df['RevLineCr'].isnull()]

LoanNr_ChkDgt                           Name           City State  \
2377       1015874006                  SALON MISSION  OVERLAND PARK    KS   
10462      1073594003      MIDDLETOWN NEWS & TOBACCO         NEWARK    DE   
30460      1229764003                     FREEZE ICE   SILVERTHORNE    CO   
40484      1313544010                GLASS CREATIONS       DONIPHAN    MO   
42828      1332844001  RICH AND CHARLIE'S RESTAURANT         EUREKA    MO   
...               ...                            ...            ...   ...   
876507     9485793000            VALUE INN SUNNYVALE      SUNNYVALE    CA   
877658     9507253004        BD HEALTH CENTERS, INC.          BARRE    VT   
880245     9557173002    MARY KAY COSM. (L) HARRISON        HOUSTON    TX   
882637     9599083008            STANLEY KASINO CLUB        STANLEY    ID   
896460     9935233005            LIGHTWAY INDUSTRIES       VALENCIA    CA   

          Zip                            Bank BankState   NAICS ApprovalDate  \
2377    66206             BANK OF BLUE VALLEY        KS       0     6-Mar-97   
10462   19702        WILMINGTON TRUST COMPANY        DE       0    28-Mar-97   
30460   80498  WELLS FARGO BANK INTERNATIONAL        NY       0    13-May-97   
40484   63935       FIRST MISSOURI STATE BANK        MO  448310     4-Jun-97   
42828   63025  PNC BANK, NATIONAL ASSOCIATION        DE       0    10-Jun-97   
...       ...                             ...       ...     ...          ...   
876507  94087   LOANS FROM OLD CLOSED LENDERS        DC       0    26-Aug-96   
877658   5641          LAKE SUNAPEE BANK, FSB        VT  621210     3-Sep-96   
880245  77033  JPMORGAN CHASE BANK NATL ASSOC        IL       0    19-Sep-96   
882637  83278  U.S. BANK NATIONAL ASSOCIATION        OH       0    30-Sep-96   
896460  91355    IOWA BUSINESS GROWTH COMPANY        IA       0    31-Jan-97   

       ApprovalFY  ...  RevLineCr  LowDoc  ChgOffDate  DisbursementDate  \
2377         1997  ...        NaN       Y         NaN         31-May-97   
10462        1997  ...        NaN       N         NaN         30-Apr-97   
30460        1997  ...        NaN       Y         NaN         30-Sep-97   
40484        1997  ...        NaN       Y         NaN         31-Oct-97   
42828        1997  ...        NaN       N         NaN         31-Dec-97   
...           ...  ...        ...     ...         ...               ...   
876507       1996  ...        NaN       N         NaN         31-Oct-96   
877658       1996  ...        NaN       N         NaN         30-Nov-96   
880245       1996  ...        NaN       Y         NaN         31-Mar-97   
882637       1996  ...        NaN       N         NaN         30-Jun-02   
896460       1997  ...        NaN       N         NaN         11-Mar-98   

        DisbursementGross  BalanceGross  MIS_Status ChgOffPrinGr  \
2377          $82,000.00         $0.00        P I F       $0.00    
10462         $70,000.00         $0.00        P I F       $0.00    
30460         $25,000.00         $0.00        P I F       $0.00    
40484         $25,076.00         $0.00        P I F       $0.00    
42828        $320,000.00         $0.00        P I F       $0.00    
...                   ...           ...         ...          ...   
876507       $639,000.00         $0.00        P I F       $0.00    
877658       $201,000.00         $0.00        P I F       $0.00    
880245         $1,567.00         $0.00          NaN       $0.00    
882637       $100,000.00         $0.00        P I F       $0.00    
896460       $232,000.00         $0.00        P I F       $0.00    

              GrAppv      SBA_Appv  
2377     $82,000.00    $65,600.00   
10462    $70,000.00    $56,000.00   
30460    $25,000.00    $20,000.00   
40484    $25,076.00    $20,061.00   
42828   $320,000.00   $240,000.00   
...              ...           ...  
876507  $639,000.00   $479,250.00   
877658  $201,000.00   $150,750.00   
880245    $2,500.00     $2,000.00   
882637  $100,000.00    $80,000.00 

In [165]:
df["MIS_Status"].isna().any()

np.True_

In [166]:
df["MIS_Status"].unique()

array(['P I F', 'CHGOFF', nan], dtype=object)

In [171]:
df[

(df["MIS_Status"] != 'P I F')
]["BalanceGross"]

6         $0.00 
36        $0.00 
43        $0.00 
58        $0.00 
60        $0.00 
           ...  
899131    $0.00 
899138    $0.00 
899143    $0.00 
899153    $0.00 
899162    $0.00 
Name: BalanceGross, Length: 159555, dtype: object

In [172]:
df['BalanceGross'].unique()

array(['$0.00 ', '$12,750.00 ', '$827,875.00 ', '$25,000.00 ',
       '$37,100.00 ', '$43,127.00 ', '$84,617.00 ', '$1,760.00 ',
       '$115,820.00 ', '$996,262.00 ', '$395,476.00 ', '$41,509.00 ',
       '$600.00 ', '$9,111.00 ', '$96,908.00 '], dtype=object)

In [189]:
df[df["MIS_Status"].isna()& df["ChgOffDate"].notnull()]

LoanNr_ChkDgt                            Name                  City  \
611        1004503009            SMITHFIELD GLASS INC             DANIELSON   
740        1005273003     GENERAL SOLAR COMPANY, INC.            SIOUX CITY   
4107       1028043007         PETIT BOIS RACQUET CLUB                BILOXI   
6569       1044653002               SON'S PARTY STORE              KENTWOOD   
9602       1066623005               ENERGY WORKS, LTD             ENGLEWOOD   
...               ...                             ...                   ...   
875570     9468613001             SHARON KAYE SWAYZER               HOUSTON   
879528     9544481010  MORLEY SUPERMKT. & VARIETY STO                MORLEY   
882390     9595923003        MARY KAY COSM. (L) HENRY             HITCHCOCK   
882845     9606423002            D & J TRANSPORT, INC  WHITE RIVER JUNCTION   
888537     9772941008  INTERNATIONAL PREHEATER COMPAN               SEATTLE   

       State    Zip                            Bank BankState  NAICS  \
611       CT   6239                             NaN       NaN      0   
740       IA  51111                             NaN       NaN      0   
4107      MS  39531                    REGIONS BANK        AL      0   
6569      MI  49508                             NaN       NaN      0   
9602      CO  80110                             NaN       NaN      0   
...      ...    ...                             ...       ...    ...   
875570    TX  77081  UNITY NATIONAL BANK OF HOUSTON        TX      0   
879528    MI  49336     MERCANTILE BANK OF MICHIGAN        MI      0   
882390    TX  77563  UNITY NATIONAL BANK OF HOUSTON        TX      0   
882845    VT   5001     WELLS FARGO BANK NATL ASSOC        SD      0   
888537    WA  98188                             NaN       NaN      0   

       ApprovalDate ApprovalFY  ...  RevLineCr  LowDoc  ChgOffDate  \
611        7-Jul-80       1980  ...          N       Y   22-May-90   
740        3-Jul-80       1980  ...          N       Y    8-Feb-90   
4107      21-Aug-80       1980  ...          N       Y   20-Dec-88   
6569      11-Sep-80       1980  ...          N       Y   27-Oct-93   
9602       8-Oct-80       1981  ...          N       Y    1-May-89   
...             ...        ...  ...        ...     ...         ...   
875570    19-Aug-96       1996  ...          N       Y   22-Mar-99   
879528    10-May-78       1978  ...          N       Y   25-Feb-91   
882390    30-Sep-96       1996  ...          N       Y   16-Feb-99   
882845     4-Oct-96       1997  ...          Y       Y    1-Mar-01   
888537    14-Mar-79       1979  ...          N       Y    9-Dec-93   

        DisbursementDate  DisbursementGross  BalanceGross  MIS_Status  \
611                  NaN                0.0        $0.00          NaN   
740                  NaN                0.0        $0.00          NaN   
4107            6-Mar-81                0.0        $0.00          NaN   
6569                 NaN                0.0        $0.00          NaN   
9602                 NaN                0.0        $0.00          NaN   
...                  ...                ...           ...         ...   
875570         30-Sep-97             3000.0        $0.00          NaN   
879528               NaN                0.0        $0.00          NaN   
882390         31-Jan-98             3000.0        $0.00          NaN   
882845         31-Dec-96              484.0        $0.00          NaN   
888537               NaN                0.0        $0.00          NaN   

        ChgOffPrinGr    GrAppv      SBA_Appv  
611     $144,461.00   300000.0  $300,000.00   
740     $142,666.00   350000.0  $350,000.00   
4107          $0.00   499200.0  $449,280.00   
6569     $18,775.00    20000.0   $20,000.00   
9602    $133,365.00   150000.0  $150,000.00   
...              ...       ...           ...  
875570    $1,961.00     3000.0    $2,400.00   
879528   $89,611.00   100000.0   $90,000.00   
882390      $407.00     3000.0    $2,400.00   
882845

In [190]:
df[df["MIS_Status"].isna()& df["ChgOffDate"].isna()]

LoanNr_ChkDgt                            Name              City State  \
343        1002585010                 Chinaberry, LTD  Saratoga Springs    NY   
738        1005264007  AMERICAN CERTIFIED HOME INSPEC             BOISE    ID   
833        1005845000                SPROUT KIDS INC.        HUNTINGTON    NY   
897        1006205009  CHINWE NDUKA MD FAMILY PRACTIC           EL PASO    TX   
1051       1007285002  Alamo Protective Services, Inc       SAN ANTONIO    TX   
...               ...                             ...               ...   ...   
892340     9855003006                          PATHCO           ALAMEDA    CA   
893791     9879293006               YAMA SEAFOOD L.A.       SAN GABRIEL    CA   
894290     9889203005          SKY PILOT RESTORATIONS          PORTLAND    OR   
896318     9932493007                   SMITH, CURTIS          RICHMOND    VA   
898182     9974823008         ROUND ROCK CHIROPRACTIC        ROUND ROCK    TX   

          Zip                            Bank BankState   NAICS ApprovalDate  \
343     12866    KEYBANK NATIONAL ASSOCIATION        OH       0    18-May-98   
738     83703     WELLS FARGO BANK NATL ASSOC        SD       0     3-Mar-97   
833     11743  JPMORGAN CHASE BANK NATL ASSOC        IL  424330     3-Mar-03   
897     79905  JPMORGAN CHASE BANK NATL ASSOC        IL  621111     9-Apr-03   
1051    78216     WELLS FARGO BANK NATL ASSOC        SD  561613    23-Jul-03   
...       ...                             ...       ...     ...          ...   
892340  94605     WELLS FARGO BANK NATL ASSOC        SD       0    19-Dec-96   
893791  91776     WELLS FARGO BANK NATL ASSOC        SD       0     6-Jan-97   
894290  97206     WELLS FARGO BANK NATL ASSOC        SD       0    10-Jan-97   
896318  23225            FIRST COMMUNITY BANK        VA       0    30-Jan-97   
898182  78681                    COMPASS BANK        AL  621310    19-Feb-97   

       ApprovalFY  ...  RevLineCr  LowDoc  ChgOffDate  DisbursementDate  \
343          1998  ...          N       Y         NaN         30-Jun-98   
738          1997  ...          N       Y         NaN         30-Apr-00   
833          2003  ...          N       Y         NaN         31-Mar-03   
897          2003  ...          N       Y         NaN         30-Apr-03   
1051         2003  ...          N       Y         NaN         31-Aug-03   
...           ...  ...        ...     ...         ...               ...   
892340       1997  ...          N       Y         NaN               NaN   
893791       1997  ...          N       Y         NaN         30-Apr-98   
894290       1997  ...          N       Y         NaN         31-Jul-97   
896318       1997  ...          N       Y         NaN         13-Feb-97   
898182       1997  ...          N       Y         NaN         29-Jan-99   

        DisbursementGross  BalanceGross  MIS_Status ChgOffPrinGr    GrAppv  \
343                 474.0        $0.00          NaN       $0.00    30000.0   
738                2585.0        $0.00          NaN       $0.00    10000.0   
833                1276.0        $0.00          NaN       $0.00    25000.0   
897                 803.0        $0.00          NaN       $0.00    40000.0   
1051               2785.0        $0.00          NaN       $0.00    20000.0   
...                   ...           ...         ...          ...       ...   
892340                0.0        $0.00          NaN       $0.00    50000.0   
893791              373.0        $0.00          NaN       $0.00    25000.0   
894290               20.0        $0.00          NaN       $0.00    10000.0   
896318             3500.0        $0.00          NaN       $0.00     3500.0   
898182                0.0        $0.00          NaN       $0.00   105000.0   

           SBA_Appv  
343     $15,000.00   
738      $5,000.00   
833     $12,500.00   
897     $20,000.00   
1051    $10,000.00   
...             ...  
892340  $25,000.00   
893791  $12,500.00   
894290   $5,000.00   
896318   $2,800.

si chgoffdate notnull == mis_status -> 0
si chgoffdate isnull = mis_status -> 1



In [183]:
df[(df["MIS_Status"] != 'P I F') & (df["MIS_Status"] != 'CHGOFF') & df["ChgOffDate"].notnull()][["ChgOffDate"]]

ChgOffDate
611     22-May-90
740      8-Feb-90
4107    20-Dec-88
6569    27-Oct-93
9602     1-May-89
...           ...
875570  22-Mar-99
879528  25-Feb-91
882390  16-Feb-99
882845   1-Mar-01
888537   9-Dec-93

[261 rows x 1 columns]

In [187]:
df[(df["MIS_Status"] == 'P I F') & df["ChgOffDate"].isnull() & (df["ChgOffPrinGr"].notnull())]

LoanNr_ChkDgt                            Name             City State  \
0          1000014003                  ABC HOBBYCRAFT       EVANSVILLE    IN   
1          1000024006     LANDMARK BAR & GRILLE (THE)        NEW PARIS    IN   
2          1000034009           WHITLOCK DDS, TODD M.      BLOOMINGTON    IN   
3          1000044001   BIG BUCKS PAWN & JEWELRY, LLC     BROKEN ARROW    OK   
4          1000054004     ANASTASIA CONFECTIONS, INC.          ORLANDO    FL   
...               ...                             ...              ...   ...   
899158     9995563001  SHADES WINDOW TINTING AUTO ALA           IRVING    TX   
899159     9995573004                    FABRIC FARMS  UPPER ARLINGTON    OH   
899160     9995603000                    FABRIC FARMS         COLUMBUS    OH   
899161     9995613003    RADCO MANUFACTURING CO.,INC.      SANTA MARIA    CA   
899163     9996003010  PACIFIC TRADEWINDS FAN & LIGHT           KAILUA    HI   

          Zip                            Bank BankState   NAICS ApprovalDate  \
0       47711                FIFTH THIRD BANK        OH  451120    28-Feb-97   
1       46526                 1ST SOURCE BANK        IN  722410    28-Feb-97   
2       47401         GRANT COUNTY STATE BANK        IN  621210    28-Feb-97   
3       74012   1ST NATL BK & TR CO OF BROKEN        OK       0    28-Feb-97   
4       32801         FLORIDA BUS. DEVEL CORP        FL       0    28-Feb-97   
...       ...                             ...       ...     ...          ...   
899158  75062   LOANS FROM OLD CLOSED LENDERS        DC       0    27-Feb-97   
899159  43221  JPMORGAN CHASE BANK NATL ASSOC        IL  451120    27-Feb-97   
899160  43221  JPMORGAN CHASE BANK NATL ASSOC        IL  451130    27-Feb-97   
899161  93455  RABOBANK, NATIONAL ASSOCIATION        CA  332321    27-Feb-97   
899163  96734            CENTRAL PACIFIC BANK        HI       0    27-Feb-97   

       ApprovalFY  ...  RevLineCr  LowDoc  ChgOffDate  DisbursementDate  \
0            1997  ...          N       Y         NaN         28-Feb-99   
1            1997  ...          N       Y         NaN         31-May-97   
2            1997  ...          N       N         NaN         31-Dec-97   
3            1997  ...          N       Y         NaN         30-Jun-97   
4            1997  ...          N       N         NaN         14-May-97   
...           ...  ...        ...     ...         ...               ...   
899158       1997  ...          N       Y         NaN         30-Jun-97   
899159       1997  ...          N       Y         NaN         30-Sep-97   
899160       1997  ...          Y       Y         NaN         31-Oct-97   
899161       1997  ...          N       N         NaN         30-Sep-97   
899163       1997  ...          N       Y         NaN         31-May-97   

        DisbursementGross  BalanceGross  MIS_Status ChgOffPrinGr    GrAppv  \
0                 60000.0        $0.00        P I F       $0.00    60000.0   
1                 40000.0        $0.00        P I F       $0.00    40000.0   
2                287000.0        $0.00        P I F       $0.00   287000.0   
3                 35000.0        $0.00        P I F       $0.00    35000.0   
4                229000.0        $0.00        P I F       $0.00   229000.0   
...                   ...           ...         ...          ...       ...   
899158            79000.0        $0.00        P I F       $0.00    79000.0   
899159            70000.0        $0.00        P I F       $0.00    70000.0   
899160            85000.0        $0.00        P I F       $0.00    85000.0   
899161           300000.0        $0.00        P I F       $0.00   300000.0   
899163            30000.0        $0.00        P I F       $0.00    30000.0   

            SBA_Appv  
0        $48,000.00   
1        $32,000.00   
2       $215,250.00   
3        $28,000.00   
4       $229,000.00   
...              ...  
899158   $63,200.00   
899159   $56,000.00   
899160   $42,500.00   
899161  $225,000.0

In [186]:
df[(df["MIS_Status"] == 'P I F') & (df["ChgOffDate"].notnull()) & (df["ChgOffPrinGr"].notnull())]

LoanNr_ChkDgt                            Name  \
558        1004176005                GREENLIFE MOBILE   
850        1005925000  GILLIAN'S DELICATESSEN & CATER   
853        1005936010  TRICAS MANAGEMENT SERVICES LLC   
861        1005986003  CAPITAL EXPLOSION GRAPHIC DESI   
866        1006016007   C-BARB AND CONLEY PROMOTIONAL   
...               ...                             ...   
895467     9913513005      INFINITE TECH. CORPORATION   
895669     9918003004           ANRO INSULATION, INC.   
897114     9950351001                        CLUB 191   
897735     9963813010   J.T. AUTOMOTIVE SERVICES, LLC   
898966     9991623009       WHEELER INVESTMENTS, INC.   

                             City State    Zip  \
558                      SAN JOSE    CA  95132   
850                   FLORAL PARK    NY  11001   
853                        DALLAS    TX  75218   
861     UPPER MARLBORO (MARLBORO)    MD  20774   
866                       HOUSTON    TX  77057   
...                           ...   ...    ...   
895467                 RICHARDSON    TX  75080   
895669                   STAFFORD    VA  22554   
897114                 PORTSMOUTH    IA  51565   
897735                SOUTHINGTON    CT   6489   
898966                   KENNESAW    GA  30144   

                                  Bank BankState   NAICS ApprovalDate  \
558                          BBCN BANK        CA  561730     7-Feb-06   
850     JPMORGAN CHASE BANK NATL ASSOC        IL  722320    26-Mar-03   
853        WELLS FARGO BANK NATL ASSOC        SD  621511     8-Feb-06   
861                          BBCN BANK        CA  541430     8-Feb-06   
866                          BBCN BANK        CA  421450     8-Feb-06   
...                                ...       ...     ...          ...   
895467                LEGACYTEXAS BANK        TX  541512    23-Jan-97   
895669   MANUFACTURERS & TRADERS TR CO        MD  235420    24-Jan-97   
897114                             NaN       NaN       0    24-Jul-78   
897735   TD BANK, NATIONAL ASSOCIATION        DE       0    13-Feb-97   
898966              BRANCH BK. & TR CO        GA  445310    26-Feb-97   

       ApprovalFY  ...  RevLineCr  LowDoc  ChgOffDate  DisbursementDate  \
558          2006  ...          N       Y   19-Oct-09         28-Feb-06   
850          2003  ...          N       Y   29-Mar-06         31-Mar-03   
853          2006  ...          Y       N    2-Feb-09         31-Mar-06   
861          2006  ...          N       Y   14-Aug-12         28-Feb-06   
866          2006  ...          N       Y    1-Jun-09         28-Feb-06   
...           ...  ...        ...     ...         ...               ...   
895467       1997  ...          N       N   29-Dec-11         31-May-97   
895669       1997  ...          N       Y   22-Mar-00         19-Feb-97   
897114       1978  ...          N       Y   23-Nov-88          6-Dec-78   
897735       1997  ...          N       Y   26-Apr-06         31-May-97   
898966       1997  ...          N       Y    9-Oct-98         31-Mar-97   

        DisbursementGross  BalanceGross  MIS_Status ChgOffPrinGr    GrAppv  \
558                5000.0        $0.00        P I F   $3,330.00     5000.0   
850               15000.0        $0.00        P I F  $10,270.00    15000.0   
853              165553.0        $0.00        P I F  $97,486.00   100000.0   
861               10000.0        $0.00        P I F   $2,310.00    10000.0   
866                5000.0        $0.00        P I F   $3,783.00     5000.0   
...                   ...           ...         ...          ...       ...   
895467           441000.0        $0.00        P I F   $6,462.00   441000.0   
895669            50000.0        $0.00        P I F  $46,165.00    50000.0   
897114            65000.0        $0.00        P I F  $20,878.00    65000.0   
897735            60000.0        $0.00        P I F  $17,236.00    60000.0   
898966            45000.0        $0.00        P I F  $30,002.00    45000.0   

       